File: projects/callisto/src/main/java/io/fairspace/callisto/config/RabbitMqConfig.java
package io.fairspace.callisto.config;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import org.springframework.amqp.core.Binding;
import org.springframework.amqp.core.BindingBuilder;
import org.springframework.amqp.core.Queue;
import org.springframework.amqp.core.TopicExchange;
import org.springframework.amqp.rabbit.connection.ConnectionFactory;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
@ConditionalOnProperty("app.rabbitmq.enabled")
public class RabbitMqConfig {
    @Bean
    RabbitTemplate rabbitTemplate(ConnectionFactory connectionFactory) {
        RabbitTemplate rabbitTemplate = new RabbitTemplate(connectionFactory);
        rabbitTemplate.setMessageConverter(producerJackson2MessageConverter());
        return rabbitTemplate;
    }

    @Bean
    Jackson2JsonMessageConverter producerJackson2MessageConverter() {
        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
        objectMapper.registerModule(new JavaTimeModule());
        return new Jackson2JsonMessageConverter(objectMapper);
    }

    @Bean
    TopicExchange collectionsExchange(RabbitMqProperties properties) {
        return new TopicExchange(properties.getTopology().getCollections().getExchange());
    }

    @Bean
    Queue permissionsAddQueue(RabbitMqProperties properties) {
        return new Queue(properties.getTopology().getCollections().getQueues().get("addPermission"), true);
    }

    @Bean
    Binding storageUploadBinding(Queue permissionsAddQueue, TopicExchange collectionsExchange) {
        return BindingBuilder.bind(permissionsAddQueue).to(collectionsExchange).with(RoutingKeys.ADD_PERMISSION);
    }

    public enum RoutingKeys {
        ADD_PERMISSION("add.permission");

        private String value;

        RoutingKeys(String v) {
            this.value = v;
        }

        public String getValue() {
            return value;
        }

        @Override
        public String toString() {
            return getValue();
        }
    }
}

File: projects/ceres/src/main/kotlin/io/fairspace/ceres/events/config/RabbitMqConfig.kt
package io.fairspace.ceres.events.config

import org.springframework.amqp.core.Binding
import org.springframework.amqp.core.BindingBuilder
import org.springframework.amqp.core.Queue
import org.springframework.amqp.core.TopicExchange
import org.springframework.amqp.rabbit.connection.ConnectionFactory
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
@ConditionalOnProperty("app.rabbitmq.enabled")
class RabbitMqConfig(val properties: RabbitMqProperties) {
    @Bean
    fun storageExchange(): TopicExchange {
        return TopicExchange(properties.topology.storage.exchange)
    }

    @Bean
    fun storageCreateQueue(): Queue {
        return Queue(properties.topology.storage.queues["create"], true)
    }

    @Bean
    fun storageMoveQueue(): Queue {
        return Queue(properties.topology.storage.queues["move"], true)
    }

    @Bean
    fun storageCopyQueue(): Queue {
        return Queue(properties.topology.storage.queues["copy"], true)
    }

    @Bean
    fun storageDeleteQueue(): Queue {
        return Queue(properties.topology.storage.queues["delete"], true)
    }

    @Bean
    fun storageUploadBinding(storageCreateQueue: Queue, storageExchange: TopicExchange): Binding {
        return BindingBuilder.bind(storageCreateQueue).to(storageExchange).with(RoutingKeys.UPLOAD)
    }

    @Bean
    fun storageMkDirBinding(storageCreateQueue: Queue, storageExchange: TopicExchange): Binding {
        return BindingBuilder.bind(storageCreateQueue).to(storageExchange).with(RoutingKeys.MKDIR)
    }

    @Bean
    fun storageMoveBinding(storageMoveQueue: Queue, storageExchange: TopicExchange): Binding {
        return BindingBuilder.bind(storageMoveQueue).to(storageExchange).with(RoutingKeys.MOVE)
    }

    @Bean
    fun storageCopyBinding(storageCopyQueue: Queue, storageExchange: TopicExchange): Binding {
        return BindingBuilder.bind(storageCopyQueue).to(storageExchange).with(RoutingKeys.COPY)
    }

    @Bean
    fun storageDeleteBinding(storageDeleteQueue: Queue, storageExchange: TopicExchange): Binding {
        return BindingBuilder.bind(storageDeleteQueue).to(storageExchange).with(RoutingKeys.DELETE)
    }


    @Bean
    fun rabbitTemplate(connectionFactory: ConnectionFactory): RabbitTemplate {
        val rabbitTemplate = RabbitTemplate(connectionFactory)
        rabbitTemplate.messageConverter = producerJackson2MessageConverter()
        return rabbitTemplate
    }

    @Bean
    fun producerJackson2MessageConverter(): Jackson2JsonMessageConverter {
        return Jackson2JsonMessageConverter()
    }


    enum class RoutingKeys(val key: String) {
        READ("read"),
        DOWNLOAD("download"),
        UPLOAD("upload"),
        MKDIR("mkdir"),
        COPY("copy"),
        MOVE("move"),
        DELETE("delete");

        override fun toString(): String {
            return key
        }
    }
}
